# --- T2-COPYRIGHT-BEGIN ---
# t2/package/*/st/10-st-hidecursor.patch
# Copyright (C) 2025 The T2 SDE Project
# SPDX-License-Identifier: GPL-2.0 or patched project license
# --- T2-COPYRIGHT-END ---

--- st-0.9.2/x.c.vanilla	2025-07-21 11:26:52.554837115 +0200
+++ st-0.9.2/x.c	2025-07-21 11:59:27.582000000 +0200
@@ -119,6 +119,11 @@
 	Draw draw;
 	Visual *vis;
 	XSetWindowAttributes attrs;
+        /* Here, we use the term *pointer* to differentiate the cursor
+         * one sees when hovering the mouse over the terminal from, e.g.,
+         * a green rectangle where text would be entered. */
+        Cursor vpointer, bpointer; /* visible and hidden pointers */
+        int pointerisvisible;
 	int scr;
 	int isfixed; /* is fixed geometry? */
 	int depth; /* bit depth */
@@ -735,6 +740,13 @@
 void
 bmotion(XEvent *e)
 {
+        if (!xw.pointerisvisible) {
+                XDefineCursor(xw.dpy, xw.win, xw.vpointer);
+                xw.pointerisvisible = 1;
+                if (!IS_SET(MODE_MOUSEMANY))
+                        xsetpointermotion(0);
+        }
+
 	if (IS_SET(MODE_MOUSE) && !(e->xbutton.state & forcemousemod)) {
 		mousereport(e);
 		return;
@@ -1159,12 +1171,12 @@
 xinit(int cols, int rows)
 {
 	XGCValues gcvalues;
-	Cursor cursor;
 	Window parent;
 	pid_t thispid = getpid();
 	XColor xmousefg, xmousebg;
 	XWindowAttributes attr;
 	XVisualInfo vis;
+	Pixmap blankpm;
 
 	xw.scr = XDefaultScreen(xw.dpy);
 
@@ -1232,8 +1244,9 @@
 	}
 
 	/* white cursor, black outline */
-	cursor = XCreateFontCursor(xw.dpy, mouseshape);
-	XDefineCursor(xw.dpy, xw.win, cursor);
+        xw.pointerisvisible = 1;
+        xw.vpointer = XCreateFontCursor(xw.dpy, mouseshape);
+        XDefineCursor(xw.dpy, xw.win, xw.vpointer);
 
 	if (XParseColor(xw.dpy, xw.cmap, colorname[mousefg], &xmousefg) == 0) {
 		xmousefg.red   = 0xffff;
@@ -1247,7 +1260,9 @@
 		xmousebg.blue  = 0x0000;
 	}
 
-	XRecolorCursor(xw.dpy, cursor, &xmousefg, &xmousebg);
+        XRecolorCursor(xw.dpy, xw.vpointer, &xmousefg, &xmousebg);
+        blankpm = XCreateBitmapFromData(xw.dpy, xw.win, &(char){0}, 1, 1);
+        xw.bpointer = XCreatePixmapCursor(xw.dpy, blankpm, blankpm, &xmousefg, &xmousebg, 0, 0);
 
 	xw.xembed = XInternAtom(xw.dpy, "_XEMBED", False);
 	xw.wmdeletewin = XInternAtom(xw.dpy, "WM_DELETE_WINDOW", False);
@@ -1774,6 +1789,9 @@
 void
 xsetpointermotion(int set)
 {
+        if (!set && !xw.pointerisvisible)
+                return;
+
 	MODBIT(xw.attrs.event_mask, set, PointerMotionMask);
 	XChangeWindowAttributes(xw.dpy, xw.win, CWEventMask, &xw.attrs);
 }
@@ -1893,6 +1911,12 @@
 	Status status;
 	Shortcut *bp;
 
+        if (xw.pointerisvisible) {
+                XDefineCursor(xw.dpy, xw.win, xw.bpointer);
+                xsetpointermotion(1);
+                xw.pointerisvisible = 0;
+        }
+
 	if (IS_SET(MODE_KBDLOCK))
 		return;
 
